### PR TITLE
Additional tests for System.Text.Json

### PR DESCRIPTION
**Pull request by alanisaac**
Adding additional tests and removed unreachable code in order to improve code coverage for System.Text.Json, as suggested in #32341.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Hi thanks for the contribution! Btw we don't really use "draft" PRs here. Is this ready for review?

**Comment by alanisaac** [comment]
@danmosemsft 

Thanks for letting me know, I've marked it as ready for review.

**Comment by alanisaac** [comment]
I added some additional changes.

### Note 1

https://github.com/dotnet/runtime/blob/fd181c0210c263e40ce8b908bdd521d8f3cc284e/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L291-L293

In the previous line, we cast the current instance to a `JsonDictionaryConverter<T>`, whose `ClassType` in practice is always `ClassType.Dictionary`.  Therefore, this conditional branch should be able to be removed safely.  I think this tracks with the purpose of this method of writing data for `[JsonExtensionData]`, which to my understanding always needs to be a `Dictionary<>` anyway.

However, to make sure this is safely enforced, I thought it would be good to explicitly mark `JsonDictionaryConverter<T>.ClassType` `sealed` as well, hence [this change](https://github.com/dotnet/runtime/pull/32705/files#diff-828e5b00d4301a141306731dd4531b1dR12).

### Note 2

https://github.com/dotnet/runtime/blob/fd181c0210c263e40ce8b908bdd521d8f3cc284e/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs#L68-L71

I'm fairly confident this is unreachable.  There are two calling methods:
* `SerializeAsync(Stream, object?, Type, JsonSerializerOptions?, CancellationToken)` already has a defensive check that throws if `inputType` is null.
* `SerializeAsync<TValue>(Stream, TValue, JsonSerializerOptions?, CancellationToken cancellationToken)` determines `inputType` from `typeof(TValue)`.  I tried to think through if there was a way that `typeof(TValue)` could be null through reflection, but I couldn't come up with anything -- `MakeGenericMethod` requires a non-null type parameter.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs [hunk 1]
```diff
@@ -21,6 +21,9 @@ public static partial class JsonSerializer
         /// <param name="cancellationToken">The <see cref="System.Threading.CancellationToken"/> which may be used to cancel the write operation.</param>
         public static Task SerializeAsync<TValue>(Stream utf8Json, TValue value, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
         {
+            if (utf8Json == null)
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs]: Nice catch. We should update the docs with the new exceptions:
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serializeasync?view=netcore-3.1#System_Text_Json_JsonSerializer_SerializeAsync_System_IO_Stream_System_Object_System_Type_System_Text_Json_JsonSerializerOptions_System_Threading_CancellationToken_

If you get the chance, please add the exceptions this (and other SerializeAsync overloads) throw, here:
https://github.com/dotnet/dotnet-api-docs/blob/1e6b2e6052185c6744054acc10cd06e435a4fb8b/xml/System.Text.Json/JsonSerializer.xml#L546-L563

See other overloads that have exception docs as an example:
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serialize?view=netcore-3.1#System_Text_Json_JsonSerializer_Serialize__1_System_Text_Json_Utf8JsonWriter___0_System_Text_Json_JsonSerializerOptions_

cc @carlossanlop

File: src/libraries/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 2]
```diff
@@ -12,9 +12,10 @@ namespace System.Text.Json.Serialization.Tests
     public static partial class StreamTests
     {
         [Fact]
-        public static async Task NullArgumentFail()
+        public static async Task ReadNullArgumentFail()
         {
             await Assert.ThrowsAsync<ArgumentNullException>(async () => await JsonSerializer.DeserializeAsync<string>((Stream)null));
+            await Assert.ThrowsAsync<ArgumentNullException>(async () => await JsonSerializer.DeserializeAsync((Stream)null, (Type)null));
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: While we are doing this, let's add all permutations. For example: have the stream parameter be null, but the type parameter be non-null.

File: src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 3]
```diff
@@ -899,6 +899,60 @@ public static void TestSingleStringsMultiSegment()
...
+        public static void TestMultiSegmentStringConversionToDateTime()
+        {
+            string jsonString = "\"1997-07-16\"";
+            string expectedString = "1997-07-16";
+
+            byte[] dataUtf8 = Encoding.UTF8.GetBytes(jsonString);
+
+            ReadOnlySequence<byte> sequence = JsonTestHelper.CreateSegments(dataUtf8);
+
+            var json = new Utf8JsonReader(sequence, isFinalBlock: false, state: default);
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: nit: Consider looping through the entire sequence and creating slices of arbitrary lengths (and re-entering the reader with "partial data"). See other "single value" tests like: `TestSingleStringsMultiSegment`

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 4]
```diff
@@ -290,45 +290,27 @@ internal bool TryWriteDataExtensionProperty(Utf8JsonWriter writer, T value, Json
             bool success;
             JsonDictionaryConverter<T> dictionaryConverter = (JsonDictionaryConverter<T>)this;
 
-            if (ClassType == ClassType.Value)
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs]: I am not sure if it is OK to remove this (it's possible that it is OK to do, but is surprising to me). What happens when the user defines and register their own `JsonConverter<T>` for dictionaries?

@steveharter, @layomia - please review this when you get a chance. I suspect this branch was needed for something (which means we need to provide the test case).

Comment (alanisaac) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs]: I touched on this in Note 1 in the above comments, but I'm not sure it's possible.  The line immediately above this one casts the current instance to a `JsonDictionaryConverter<T>` which is marked `internal`, so my assumption was users wouldn't be able to inherit it (unless there's something I'm not thinking of or aware of?)

I can certainly leave this in there if that's the call, but I think then testing and getting coverage on these lines would be difficult without introducing an `InternalsVisibleTo` relationship (or similar internals-testing-pattern) between the src and test project.  That or adding an inherited class in the src project for the purposes of testing.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs (hunk 1 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs]: Will do, thanks for the pointers to the docs code and example.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.ReadTests.cs (hunk 2 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: fixed in 7a6e0cbc

File: src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs (hunk 3 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: Tried something out with 823f147, let me know if that covers what you were thinking.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonDictionaryConverter.cs [hunk 5]
```diff
@@ -9,7 +9,7 @@ namespace System.Text.Json.Serialization
     /// </summary>
     internal abstract class JsonDictionaryConverter<T> : JsonResumableConverter<T>
     {
-        internal override ClassType ClassType => ClassType.Dictionary;
+        internal sealed override ClassType ClassType => ClassType.Dictionary;
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonDictionaryConverter.cs]: We should seal the overrides in `JsonCollectionConverter` and `JsonObjectConverter` as well.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs]: There is a bug with writing extension data if there is a custom converter for the extension property: https://github.com/dotnet/runtime/issues/32903. This is the reason why the L291 is not problematic (`InvalidCastException`) - we throw an NRE up stack.

We'll likely need this branch when we fix that bug, so for this PR we should revert this change.

Comment (alanisaac) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs]: reverted in 256de92

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonDictionaryConverter.cs (hunk 5 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonDictionaryConverter.cs]: fixed in 02ab96b

**Review by layomia — APPROVED** [review]
Thanks, @alanisaac!

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs (hunk 1 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs]: Created issue https://github.com/dotnet/dotnet-api-docs/issues/3961

File: src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 6]
```diff
@@ -899,6 +899,82 @@ public static void TestSingleStringsMultiSegment()
...
+
+        private static void ReadDateTimeHelper(ref Utf8JsonReader jsonReader, DateTime expectedValue, long expectedTokenLength)
+        {
+            while (jsonReader.Read())
+            {
+                if (jsonReader.TokenType == JsonTokenType.String)
+                {
+                    long tokenLength = jsonReader.HasValueSequence ? jsonReader.ValueSequence.Length : jsonReader.ValueSpan.Length;
+                    Assert.Equal(expectedTokenLength, tokenLength);
+                    Assert.Equal(expectedValue, jsonReader.GetDateTime());
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: nit: We missed the asserts for `TryGetDateTime` and `TryGetDateTimeOffset` APIs.

**Pull request merged**